espnet2/enh/separator/tfgridnet_separator.py

```diff
@@ -0,0 +1,402 @@
+
+
+class GridNetBlock(nn.Module):
+    def __getitem__(self, key):
+        return getattr(self, key)
+
+    def __init__(
+        self,
+        emb_dim,
+        emb_ks,
+        emb_hs,
+        n_freqs,
+        hidden_channels,
+        n_head=4,
+        approx_qk_dim=512,
+        activation="prelu",
+        eps=1e-5,
+    ):
+        super(GridNetBlock, self).__init__()
+
+        in_channels = emb_dim * emb_ks
+
+        self.intra_norm = LayerNormalization4D(emb_dim, eps=eps)
+        self.intra_rnn = nn.LSTM(
+            in_channels, hidden_channels, 1, batch_first=True, bidirectional=True
+        )
+        self.intra_linear = nn.ConvTranspose1d(
+            hidden_channels * 2, emb_dim, emb_ks, stride=emb_hs
+        )
+
+        self.inter_norm = LayerNormalization4D(emb_dim, eps=eps)
+        self.inter_rnn = nn.LSTM(
+            in_channels, hidden_channels, 1, batch_first=True, bidirectional=True
+        )
+        self.inter_linear = nn.ConvTranspose1d(
+            hidden_channels * 2, emb_dim, emb_ks, stride=emb_hs
+        )
+
+        E = math.ceil(
+            approx_qk_dim * 1.0 / n_freqs
+        )  # approx_qk_dim is only approximate
+        assert emb_dim % n_head == 0
+        for ii in range(n_head):
+            self.add_module(
+                "attn_conv_Q_%d" % ii,
+                nn.Sequential(
+                    nn.Conv2d(emb_dim, E, 1),
+                    get_layer(activation)(),
+                    LayerNormalization4DCF((E, n_freqs), eps=eps),
+                ),
+            )
+            self.add_module(
+                "attn_conv_K_%d" % ii,
+                nn.Sequential(
+                    nn.Conv2d(emb_dim, E, 1),
+                    get_layer(activation)(),
+                    LayerNormalization4DCF((E, n_freqs), eps=eps),
+                ),
+            )
+            self.add_module(
+                "attn_conv_V_%d" % ii,
+                nn.Sequential(
+                    nn.Conv2d(emb_dim, emb_dim // n_head, 1),
+                    get_layer(activation)(),
+                    LayerNormalization4DCF((emb_dim // n_head, n_freqs), eps=eps),
+                ),
+            )
+        self.add_module(
+            "attn_concat_proj",
+            nn.Sequential(
+                nn.Conv2d(emb_dim, emb_dim, 1),
+                get_layer(activation)(),
+                LayerNormalization4DCF((emb_dim, n_freqs), eps=eps),
+            ),
+        )
+
+        self.emb_dim = emb_dim
+        self.emb_ks = emb_ks
+        self.emb_hs = emb_hs
+        self.n_head = n_head
+
+    def forward(self, x):
+        """
+        x: [B, C, T, Q]
+        out: [B, C, T, Q]
+        """
+        B, C, old_T, old_Q = x.shape
+        T = (
+            math.ceil((old_T - self.emb_ks) * 1.0 / self.emb_hs) * self.emb_hs
+            + self.emb_ks
+        )
+        Q = (
+            math.ceil((old_Q - self.emb_ks) * 1.0 / self.emb_hs) * self.emb_hs
+            + self.emb_ks
+        )
```

Comment:
`* 1.0` can be removed as the output of division `/` is a floating-point number in Python3.